### ISSUE TITLE
Lock digest differs when performing dependency build on chart in Helm 2 and then Helm 3

### ISSUE THREAD

**Issue by taylorsilva** [helm/helm#7233]
When doing `helm dependency build` with the Helm v3 CLI on a chart using the Helm `v1` API, the following error occurs:

```
┌─[/tmp/test-chart]
└─▪ helm dependency build
Error: Chart.lock is out of sync with Chart.yaml
```

#### Expected Behaviour
Based on the [Helm v2 to v3 FAQ](https://v3.helm.sh/docs/faq/#consolidation-of-requirements-yaml-into-chart-yaml), Helm v3 should behave the same as Helm v2 with a chart using the Helm `v1` API.

### Steps To Reproduce

1. Create a new chart using the Helm v2 CLI

```
┌─[/tmp]
└─▪ helm2 create test-chart
Creating test-chart
```

2. Add a `requirements.yaml` to the chart and add one or more dependencies. I used the following:

```yaml
dependencies:
- name: postgresql
  version: 6.5.5
  repository: https://kubernetes-charts.storage.googleapis.com/
  condition: postgresql.enabled
```

3. Run `dependency build` using the Helm v2 and v3 CLI, observing the different outputs.
Helm v2
```
┌─[/tmp/test-chart]
└─▪ helm2 dependency build
Hang tight while we grab the latest from your chart repositories...
...Unable to get an update from the "local" chart repository (http://127.0.0.1:8879/charts):
	Get http://127.0.0.1:8879/charts/index.yaml: dial tcp 127.0.0.1:8879: connect: connection refused
...Successfully got an update from the "stable" chart repository
Update Complete.
Saving 1 charts
Downloading postgresql from repo https://kubernetes-charts.storage.googleapis.com/
Deleting outdated charts
```
Helm v3
```
┌─[/tmp/test-chart]
└─▪ helm dependency build
Error: Chart.lock is out of sync with Chart.yaml
```

### CLI Versions

```
└─▪ helm2 version
Client: &version.Version{SemVer:"v2.16.1", GitCommit:"bbdfe5e7803a12bbdf97e94cd847859890cf4050", GitTreeState:"clean"}
└─▪ helm version
version.BuildInfo{Version:"v3.0.0", GitCommit:"e29ce2a54e96cd02ccfce88bee4f58bb6e2a28b6", GitTreeState:"clean", GoVersion:"go1.13.4"}
```

**Comment by zwwhdls** [comment]
Hi, would you use helm 3.0.1 to reproduce it? Helm 3.0.1 has solved a same problem: https://github.com/helm/helm/pull/7119.

If you can reproduced this by Helm 3.0.1, can you provide a more detailed description for reproduced it ?

**Comment by taylorsilva** [comment]
I was able to reproduce this issue with helm `3.0.1`

```
┌─[/tmp/test-chart]
└─▪ helm version
version.BuildInfo{Version:"v3.0.1", GitCommit:"7c22ef9ce89e0ebeb7125ba2ebf7d421f3e82ffa", GitTreeState:"clean", GoVersion:"go1.13.4"}
┌─[/tmp/test-chart]
└─▪ helm dependency build
Error: Chart.lock is out of sync with Chart.yaml
```

There is no `Chart.lock` file present before of after `helm dependency build` is executed.

**Comment by hickeyma** [comment]
@taylorsilva This might also be affected by issue fixed in #7009. Do you mind checking against the latest Helm 3 code in master to see if the issue is fixed? Can you also remove any `Chart.lock` file and check that your dependencies are defined in `requirements.yaml` only.

**Comment by taylorsilva** [comment]
I was able to reproduce the issue using the latest code on master.

```
/t/test-chart $ /tmp/helm version
version.BuildInfo{Version:"v3.0", GitCommit:"0fd76b2a2bca5b3febf6488c860a7c6502c70086", GitTreeState:"clean", GoVersion:"go1.13.4"}

/t/test-chart $ ll
total 32
-rw-r--r--  1 pivotal  wheel   106B Dec 16 09:39 Chart.yaml
drwxr-xr-x  3 pivotal  wheel    96B Dec 16 09:39 charts
-rw-r--r--  1 pivotal  wheel   240B Dec 16 09:39 requirements.lock
-rw-r--r--  1 pivotal  wheel   146B Dec 16 09:39 requirements.yaml
drwxr-xr-x  9 pivotal  wheel   288B Dec 16 09:39 templates
-rw-r--r--  1 pivotal  wheel   1.5K Dec 16 09:39 values.yaml

/t/test-chart $ /tmp/helm dependency build
Error: Chart.lock is out of sync with Chart.yaml

/t/test-chart $ ll
total 32
-rw-r--r--  1 pivotal  wheel   106B Dec 16 09:39 Chart.yaml
drwxr-xr-x  3 pivotal  wheel    96B Dec 16 09:39 charts
-rw-r--r--  1 pivotal  wheel   240B Dec 16 09:39 requirements.lock
-rw-r--r--  1 pivotal  wheel   146B Dec 16 09:39 requirements.yaml
drwxr-xr-x  9 pivotal  wheel   288B Dec 16 09:39 templates
-rw-r--r--  1 pivotal  wheel   1.5K Dec 16 09:39 values.yaml
```

Sanity checks:
- Still works using Helm v2
```
/t/test-chart $ helm2 dependency build
Hang tight while we grab the latest from your chart repositories...
...Successfully got an update from the "harbor" chart repository
...Successfully got an update from the "concourse" chart repository
...Successfully got an update from the "stable" chart repository
Update Complete.
Saving 1 charts
Downloading postgresql from repo https://kubernetes-charts.storage.googleapis.com/
Deleting outdated charts
```
- There's no `Chart.lock` file before or after running any of the `dependency build` commands
- Contents of `requirements.yaml` has not changed before or after running any of the `dependency build` commands

**Comment by karuppiah7890** [comment]
@taylorsilva Thanks for helping in reproducing the issue. I'm able to see the bug. @hickeyma I think we need to label this as `bug`.

**Comment by waveywaves** [comment]
I was able to reproduce this as well

**Comment by karuppiah7890** [comment]
@waveywaves you wanna work on this? 😄

**Comment by hickeyma** [comment]
Sorry for the delay in replying but got side tracked before I could reply with my investigations.

@taylorsilva Thanks for describing how to reproduce in https://github.com/helm/helm/issues/7233#issuecomment-566092411.

I have been looking at the issue and trying to work out the best way to handle it. As it stands the logic is working as expected. The issue is happening because when apiVersion v1 chart dependencies are built with Helm 2 and then built with Helm 3, the lock digests differ.  This is to be expected as different code builds the digest in the Helm 2 and 3 cases.

Should the logic remain as is or should it realign in code. @waveywaves I am not sure if you were working on this but I hope it is ok that I pushed PR #7261 as a possible solution that I had ready.

**Comment by TBBle** [comment]
I assume the root cause for this issue was the change in hashing mechanism introduced to fix #2731?

Perhaps the API-stablest approach would be to teach `HashReq` to generate the right kind of hash depending on the chart version?

However, that would reintroduce #2731 into APIVersionV1 charts in Helm v3, so it might be better to treat those older hashes as legacy. They would be accepted during `helm dep build` or similar validations, but never *generated* by Helm v3.

That in-turn might introduce issues for Helm v2 validating APIVersionV1 charts generated by Helm v3. Which would be a real shame, getting the fix for #2731 was a driver for my site upgrading to Helm v3 pretty much the week it was released.

Is `helm dep build` the only thing that actually validates this hash? We locally commit our charts as well as the lockfile, so we never use `helm dep build`, and so would never have noticed this issue ourselves, but it's only because we have `skipDependenciesBuild` turned off in our skaffold.yaml for historical reasons.

**Comment by technosophos** [comment]
One thing to consider here would be to emit a warning when the hash verification failed, but the versions are different. Something like:

 > WARNING: Chart $NAME was generated with an older version of Helm and its lock file cannot be verified. Repackage it with Helm 3 to fix this.

I would suggest adding a `strict` flag of some sort to allow security-minded individuals a way of preventing this. However, with signed charts I can't think of any real security issue this introduces. (With unsigned charts, though, it is admittedly dangerous)

While this is ugly, it doesn't hide what is happening, whereas the other solutions i have seen attempt to silently work around the hash mismatch by regenerating, ignoring, or updating.

- Regenerating is safer if and only if we can do it without altering any "meaningful" data. But I haven't yet seen a good way of doing this.
- Ignoring is always bad, and opens security holes
- Updating basically ignores the entire intention of having a lock file in the first place, since the build is not reproducible, but I think it would be fine to add a flag that let people CHOOSE to force an update.

The lockfile does have a _specific_ version of a package in it (as opposed to a SemVer pattern). So we could regenerate just the hashes by reading a lock, fetching the packages, and re-hashing. 

Here's the Lock.Dependency struct:

```go
type Dependency struct {
	// Name is the name of the dependency.
	//
	// This must mach the name in the dependency's Chart.yaml.
	Name string `json:"name"`
	// Version is the version (range) of this chart.
	//
	// A lock file will always produce a single version, while a dependency
	// may contain a semantic version range.
	Version string `json:"version,omitempty"`
	// The URL to the repository.
	//
	// Appending `index.yaml` to this string should result in a URL that can be
	// used to fetch the repository index.
	Repository string `json:"repository"`
	// A yaml path that resolves to a boolean, used for enabling/disabling charts (e.g. subchart1.enabled )
	Condition string `json:"condition,omitempty"`
	// Tags can be used to group charts for enabling/disabling together
	Tags []string `json:"tags,omitempty"`
	// Enabled bool determines if chart should be loaded
	Enabled bool `json:"enabled,omitempty"`
	// ImportValues holds the mapping of source values to parent key to be imported. Each item can be a
	// string or pair of child/parent sublist items.
	ImportValues []interface{} `json:"import-values,omitempty"`
	// Alias usable alias to be used for the chart
	Alias string `json:"alias,omitempty"`
}

```

It is possible to parse into that struct, then get the exact dependencies. Then fetch them and regenerate just the hashes. Again, it is not _secure_ unless charts are all signed. But it is one way to try to reproduce an existing app without having to call `Update` on the manager.

**Comment by TBBle** [comment]
Would you even need to download everything again? It just needs to update the digest against the right structure based on the input. The algorithm didn't change, it just went from hashing the requirements, to hashing the requirements+lock (#5874 + #6736). I don't see any recursion or walking for sub-charts involved here, and my _recollection_ is that the `chart.Dependency` structure didn't change particularly in the Helm v2->v3 change, when I went looking previously.

**Comment by hickeyma** [comment]
Fixed with merge of #7261.
